proposals/0000-shared-storage-property-wrapper.md

```diff
@@ -0,0 +1,419 @@
+
+```swift
+@propertyWrapper
+struct Wrapper<Value> {
+  shared let storage: SharedStorage
+  var wrappedValue: Value { ... }
+  
+  // a plain struct, with properties that will be shared across wrappers
+  struct SharedStorage { ... }
+}
+```
+
+## Detailed design
+
+The storage is declared using the new `shared` property attribute inside a Property Wrapper declaration. This property will be initialized and stored globally by the compiler, while remaining accessible to the property wrapper instance like any other private property defined in the wrapper type. 
+
+In the next example, the `RangeStorage` struct will be used for the `Clamped` wrapper. 
+
+```swift
+struct RangeStorage<Value: Comparable> {
+  let range: ClosedRange<Value>
+  init(_ range: ClosedRange<Value>) { 
+    self.range = range
+  } 
+  
+  func clamp(_ value: Value) -> Value {}
+}
+
+@propertyWrapper
+struct Clamped<Value: Comparable> {
+  shared let storage: RangeStorage<Value>
+  private var value: Value
+  
+  init(wrappedValue: Value, shared: RangeStorage<Value>) {
+    self.value = shared.clamp(wrappedValue)
+  }
+  
+  var wrappedValue: Value {
+    get { value }
+    set {
+      // `storage` is available like any other property
+      value = storage.clamp(newValue) 
+    }
+  }
+}  
+```
+
+And at the point of use, the compiler will make sure `RangeStorage` is initialized once for each wrapper application, and stored at a scope outside of the instances. Later on, when multiple instances of `Hud` are initialized, they'll all be given access to the same `$shared` properties. Suggestions on how to name the `$shared` property would be much appreciated. 🙂
+
+```swift
+struct Hud {
+  @Clamped(.init(0...14)) var bar = 5
+  @Clamped(.init(1...9)) var foo = 1
+}
+
+var hud1 = Hud()
+var hud2 = Hud()
+
+// desugars to
+
+struct Hud {
+  shared let bar$shared = RangeStorage(0...14)
+  shared let foo$shared = RangeStorage(1...9)
+  
+  var bar = Clamped(wrappedValue: 5, shared: bar$shared)
+  var foo = Clamped(wrappedValue: 1, shared: foo$shared)
+}
+
+// both Hud's get access to the same $shared properties.
+var hud1 = Hud() 
+var hud2 = Hud() 
+```
+
+### Initialization
+
+Inside the wrapper's initializer, assigning the shared value to the `shared` property is handled by the compiler, so there's not need to explicitly do it. If the initializer itself needs access to storage properties, it should be done through the `shared` parameter. 
+
+```swift
+shared let storage: RangeStorage<Value>
+
+init(wrappedValue: Value, shared: RangeStorage<Value>) {
+  self.value = shared.clamp(wrappedValue)
+
+  // referencing `storage` inside the initializer is not okay since 
+  // it might not have been initialized yet?
```

Comment:
@hborla I think we briefly touched on this but I'm not sure if it would really be a problem to access `storage` directly. Since it would be initialized already. 🤔

Reply:
I agree - accessing the shared property directly is fine here